src/Elmish.WPF/Program.fs

```diff
@@ -1,101 +1,129 @@
-
-
-/// Starts the Elmish and WPF dispatch loops with the specified configuration.
-/// Will instantiate Application and set its MainWindow if it is not already
-/// running, and then run the specified window. This is a blocking function.
-let runWindowWithConfig config (window: Window) program =
-  initializeApplication window
-  window.Show ()
-  startElmishLoop config window program
-  Application.Current.Run window
-
-
-/// Starts the Elmish and WPF dispatch loops. Will instantiate Application and
-/// set its MainWindow if it is not already running, and then run the specified
-/// window. This is a blocking function.
-let runWindow window program =
-  runWindowWithConfig ElmConfig.Default window program
-
-
-/// Same as mkSimple, but with a signature adapted for Elmish.WPF.
-let mkSimpleWpf
-    (init: unit -> 'model)
-    (update: 'msg  -> 'model -> 'model)
-    (bindings: unit -> Binding<'model, 'msg> list) =
-  Program.mkSimple init update (fun _ _ -> bindings ())
-
-
-/// Same as mkProgram, but with a signature adapted for Elmish.WPF.
-let mkProgramWpf
-    (init: unit -> 'model * Cmd<'msg>)
-    (update: 'msg  -> 'model -> 'model * Cmd<'msg>)
-    (bindings: unit -> Binding<'model, 'msg> list) =
-  Program.mkProgram init update (fun _ _ -> bindings ())
-
-
-/// Same as mkProgramWpf, except that init and update doesn't return Cmd<'msg>
-/// directly, but instead return a CmdMsg discriminated union that is converted
-/// to Cmd<'msg> using toCmd. This means that the init and update functions
-/// return only data, and thus are easier to unit test. The CmdMsg pattern is
-/// general; this is just a trivial convenience function that automatically
-/// converts CmdMsg to Cmd<'msg> for you in inint and update
-let mkProgramWpfWithCmdMsg
-    (init: unit -> 'model * 'cmdMsg list)
-    (update: 'msg -> 'model -> 'model * 'cmdMsg list)
-    (bindings: unit -> Binding<'model, 'msg> list)
-    (toCmd: 'cmdMsg -> Cmd<'msg>) =
-  let convert (model, cmdMsgs) =
-    model, (cmdMsgs |> List.map toCmd |> Cmd.batch)
-  mkProgramWpf
-    (init >> convert)
-    (fun msg model -> update msg model |> convert)
-    bindings
-
-
-/// Traces all updates using System.Diagnostics.Debug.WriteLine.
-let withDebugTrace program =
-  program |> Program.withTrace (fun msg model ->
-    System.Diagnostics.Debug.WriteLine(sprintf "New message: %A" msg)
-    System.Diagnostics.Debug.WriteLine(sprintf "Updated state: %A" model)
-  )
+type ProgramWpf<'arg, 'model, 'msg> = private {
+  ElmishProgram: Program<'arg, 'model, 'msg, Binding<'model, 'msg> list>
+  ElmConfig: ElmConfig
+}
+
+
+[<RequireQualifiedAccess>]
+module Program =
+
+  /// Starts the Elmish dispatch loop, setting the bindings as the DataContext
+  /// for the specified FrameworkElement. Non-blocking. This is a low-level function;
+  /// for normal usage, see runWindow and runWindowWithConfig.
+  let startElmishLoop
+      (element: FrameworkElement)
+      (program: ProgramWpf<unit, 'model, 'msg>) =
```

Comment:
The `'arg` type in `Progrram` is used in traditional Elmish apps to pass an initial state to the top-level `init` function, typically from a URL parser. This doesn't apply to Elmish.WPF applications. hence the reason `'arg` is always `unit`. With `ProgramWpf` now abstracting `Program`, is it worth excluding `'arg` from the definition of `ProgramWpf`?

Reply:
I am definitely in favor of this.  Thanks for the suggestion.

Dropping the `'arg` type was also considered for `Elmish.Program` https://github.com/elmish/elmish/issues/138, but it didn't go anywhere.

Do you know of an example of a traditional Elmish/Fable app that passes initial state into the top-level `init` function?  I would like to try a refactor attempt like Don Syme did.

Reply:
I pushed a commit with this change

Reply:
Are you referring to the refactor in elmish/elmish#138? The thing is, I don't think we will even need a non-unit argument passed to `init`, will we? The only use-case I've ever seen in Elmish is when a URL is parsed during state initialization.

The [sample-react-navigation](https://github.com/elmish/sample-react-navigation) shows a typical example, specifically [here](https://github.com/elmish/sample-react-navigation/blob/master/src/app.fs#L73).

Reply:
That file was edited slightly.  Here is the [corresponding permalink](https://github.com/elmish/sample-react-navigation/blob/c5fe721b29ffcf4e75ae55d7f718495fe0d705d5/src/app.fs#L73).